File: containers/shared/json-safe-encoding.js
// JSON Safe Encoding Utilities for Da-Kraken
// Provides secure JSON handling across all language containers

/**
 * Safe JSON parsing with error handling and validation
 * @param {string} jsonString - JSON string to parse
 * @param {object} options - Parsing options
 * @returns {object} Parsed object or error info
 */
function safeJsonParse(jsonString, options = {}) {
    const { 
        maxDepth = 10, 
        maxStringLength = 10000,
        allowedTypes = ['string', 'number', 'boolean', 'object', 'array'],
        sanitizeHtml = true
    } = options;

    try {
        // Input validation
        if (typeof jsonString !== 'string') {
            throw new Error('Input must be a string');
        }

        if (jsonString.length > maxStringLength) {
            throw new Error(`JSON string exceeds maximum length of ${maxStringLength}`);
        }

        // Parse with depth checking
        const parsed = JSON.parse(jsonString, (key, value) => {
            // Check nesting depth
            const depth = key.split('.').length;
            if (depth > maxDepth) {
                throw new Error(`JSON nesting exceeds maximum depth of ${maxDepth}`);
            }

            // Sanitize strings if requested
            if (sanitizeHtml && typeof value === 'string') {
                value = sanitizeString(value);
            }

            return value;
        });

        // Validate structure
        validateJsonStructure(parsed, allowedTypes);

        return { success: true, data: parsed };

    } catch (error) {
        return { 
            success: false, 
            error: error.message,
            type: 'JSON_PARSE_ERROR'
        };
    }
}

/**
 * Safe JSON stringification with circular reference protection
 * @param {any} obj - Object to stringify
 * @param {object} options - Stringification options
 * @returns {string} Safe JSON string
 */
function safeJsonStringify(obj, options = {}) {
    const {
        maxDepth = 10,
        maxStringLength = 10000,
        replacer = null,
        space = null,
        escapeHtml = true
    } = options;

    const seen = new WeakSet();
    let depth = 0;

    const safeReplacer = (key, value) => {
        // Check depth
        if (key && depth++ > maxDepth) {
            return '[Max Depth Exceeded]';
        }

        // Handle circular references
        if (typeof value === 'object' && value !== null) {
            if (seen.has(value)) {
                return '[Circular Reference]';
            }
            seen.add(value);
        }

        // Apply custom replacer if provided
        if (replacer && typeof replacer === 'function') {
            value = replacer(key, value);
        }

        // Escape HTML in strings if requested
        if (escapeHtml && typeof value === 'string') {
            value = escapeHtmlString(value);
        }

        // Limit string length
        if (typeof value === 'string' && value.length > maxStringLength) {
            value = value.substring(0, maxStringLength) + '[Truncated]';
        }

        depth--;
        return value;
    };

    try {
        return JSON.stringify(obj, safeReplacer, space);
    } catch (error) {
        return JSON.stringify({
            error: 'JSON_STRINGIFY_ERROR',
            message: error.message,
            originalType: typeof obj
        });
    }
}

/**
 * Sanitize string content to prevent XSS and injection attacks
 * @param {string} str - String to sanitize
 * @returns {string} Sanitized string
 */
function sanitizeString(str) {
    if (typeof str !== 'string') return str;
    
    return str
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#x27;')
        .replace(/\//g, '&#x2F;')
        .replace(/\\/g, '&#x5C;')
        .replace(/\x00/g, '') // Remove null bytes
        .replace(/[\x01-\x08\x0B\x0C\x0E-\x1F\x7F]/g, ''); // Remove control characters
}

/**
 * Escape HTML characters in strings
 * @param {string} str - String to escape
 * @returns {string} Escaped string
 */
function escapeHtmlString(str) {
    if (typeof str !== 'string') return str;
    
    const htmlEscapes = {
        '&': '&amp;',
        '<': '&lt;',
        '>': '&gt;',
        '"': '&quot;',
        "'": '&#x27;',
        '/': '&#x2F;'
    };
    
    return str.replace(/[&<>"'\/]/g, char => htmlEscapes[char]);
}

/**
 * Validate JSON structure against allowed types
 * @param {any} obj - Object to validate
 * @param {array} allowedTypes - Array of allowed types
 */
function validateJsonStructure(obj, allowedTypes) {
    function validate(value, path = 'root') {
        const type = Array.isArray(value) ? 'array' : typeof value;
        
        if (!allowedTypes.includes(type)) {
            throw new Error(`Invalid type '${type}' at path '${path}'. Allowed types: ${allowedTypes.join(', ')}`);
        }

        if (type === 'object' && value !== null) {
            for (const [key, val] of Object.entries(value)) {
                validate(val, `${path}.${key}`);
            }
        } else if (type === 'array') {
            value.forEach((val, index) => {
                validate(val, `${path}[${index}]`);
            });
        }
    }

    validate(obj);
}

/**
 * Create a secure JSON API response wrapper
 * @param {any} data - Response data
 * @param {object} meta - Metadata
 * @returns {object} Formatted API response
 */
function createApiResponse(data, meta = {}) {
    const response = {
        success: true,
        timestamp: new Date().toISOString(),
        data: data,
        meta: {
            version: '1.0',
            encoding: 'utf-8',
            contentType: 'application/json',
            ...meta
        }
    };

    return safeJsonStringify(response, { escapeHtml: true });
}

/**
 * Create an error response
 * @param {string} message - Error message
 * @param {string} code - Error code
 * @param {any} details - Additional error details
 * @returns {object} Error response
 */
function createErrorResponse(message, code = 'UNKNOWN_ERROR', details = null) {
    const response = {
        success: false,
        error: {
            message: sanitizeString(message),
            code: code,
            timestamp: new Date().toISOString(),
            details: details
        }
    };

    return safeJsonStringify(response, { escapeHtml: true });
}

// Export for Node.js environments
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        safeJsonParse,
        safeJsonStringify,
        sanitizeString,
        escapeHtmlString,
        validateJsonStructure,
        createApiResponse,
        createErrorResponse
    };
}

// Global object for browser environments
if (typeof window !== 'undefined') {
    window.JsonSafeEncoding = {
        safeJsonParse,
        safeJsonStringify,
        sanitizeString,
        escapeHtmlString,
        validateJsonStructure,
        createApiResponse,
        createErrorResponse
    };
}

File: containers/shared/json_safe_encoding.py
"""
JSON Safe Encoding Utilities for Python - Da-Kraken
Provides secure JSON handling for Python containers
"""

import json
import html
import re
from typing import Any, Dict, Optional
from datetime import datetime


class JsonSafeEncoder:
    """Secure JSON encoder with safety checks and sanitization"""

    def __init__(self, max_depth: int = 10, max_string_length: int = 10000):
        self.max_depth = max_depth
        self.max_string_length = max_string_length
        self._seen_objects = set()

    def safe_json_parse(self, json_string: str, **options) -> Dict[str, Any]:
        """
        Safely parse JSON string with validation and sanitization

        Args:
            json_string: JSON string to parse
            **options: Additional parsing options

        Returns:
            Dictionary with success status and data or error info
        """
        try:
            # Input validation
            if not isinstance(json_string, str):
                raise ValueError("Input must be a string")

            if len(json_string) > self.max_string_length:
                raise ValueError(
                    f"JSON string exceeds maximum length of {self.max_string_length}"
                )

            # Parse JSON
            parsed_data = json.loads(json_string)

            # Validate and sanitize
            sanitized_data = self._sanitize_data(parsed_data, depth=0)

            return {"success": True, "data": sanitized_data}

        except (json.JSONDecodeError, ValueError) as e:
            return {"success": False, "error": str(e), "type": "JSON_PARSE_ERROR"}

    def safe_json_stringify(self, obj: Any, **options) -> str:
        """
        Safely stringify object to JSON with circular reference protection

        Args:
            obj: Object to stringify
            **options: Additional stringification options

        Returns:
            Safe JSON string
        """
        try:
            self._seen_objects.clear()

            # Configure options
            indent = options.get("indent", None)
            ensure_ascii = options.get("ensure_ascii", False)
            escape_html = options.get("escape_html", True)

            # Process object with safety checks
            safe_obj = self._make_json_safe(obj, depth=0)

            # Stringify with custom encoder
            result = json.dumps(
                safe_obj,
                indent=indent,
                ensure_ascii=ensure_ascii,
                cls=SafeJSONEncoder,
                default=self._default_serializer,
            )

            if escape_html:
                result = html.escape(result)

            return result

        except Exception as e:
            return json.dumps(
                {
                    "error": "JSON_STRINGIFY_ERROR",
                    "message": str(e),
                    "original_type": str(type(obj)),
                }
            )

    def _sanitize_data(self, data: Any, depth: int = 0) -> Any:
        """Recursively sanitize data structure"""
        if depth > self.max_depth:
            return "[Max Depth Exceeded]"

        if isinstance(data, str):
            return self.sanitize_string(data)
        elif isinstance(data, dict):
            return {
                self.sanitize_string(str(k)): self._sanitize_data(v, depth + 1)
                for k, v in data.items()
            }
        elif isinstance(data, list):
            return [self._sanitize_data(item, depth + 1) for item in data]
        else:
            return data

    def _make_json_safe(self, obj: Any, depth: int = 0) -> Any:
        """Make object safe for JSON serialization"""
        if depth > self.max_depth:
            return "[Max Depth Exceeded]"

        # Handle circular references
        obj_id = id(obj)
        if obj_id in self._seen_objects:
            return "[Circular Reference]"

        if isinstance(obj, (dict, list, tuple, set)):
            self._seen_objects.add(obj_id)

        try:
            if isinstance(obj, dict):
                result = {}
                for k, v in obj.items():
                    safe_key = str(k)[: self.max_string_length]
                    result[safe_key] = self._make_json_safe(v, depth + 1)
                return result
            elif isinstance(obj, (list, tuple)):
                return [self._make_json_safe(item, depth + 1) for item in obj]
            elif isinstance(obj, set):
                return [self._make_json_safe(item, depth + 1) for item in obj]
            elif isinstance(obj, str):
                return obj[: self.max_string_length]
            else:
                return obj
        finally:
            if isinstance(obj, (dict, list, tuple, set)):
                self._seen_objects.discard(obj_id)

    def _default_serializer(self, obj: Any) -> Any:
        """Default serializer for non-JSON-serializable objects"""
        if hasattr(obj, "isoformat"):  # datetime objects
            return obj.isoformat()
        elif hasattr(obj, "__dict__"):  # Custom objects
            return obj.__dict__
        else:
            return str(obj)

    @staticmethod
    def sanitize_string(text: str) -> str:
        """Sanitize string to prevent XSS and injection attacks"""
        if not isinstance(text, str):
            return str(text)

        # Remove control characters
        text = re.sub(r"[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]", "", text)

        # HTML escape
        text = html.escape(text, quote=True)

        return text

    def create_api_response(self, data: Any, meta: Optional[Dict] = None) -> str:
        """Create a secure API response wrapper"""
        if meta is None:
            meta = {}

        response = {
            "success": True,
            "timestamp": datetime.utcnow().isoformat(),
            "data": data,
            "meta": {
                "version": "1.0",
                "encoding": "utf-8",
                "content_type": "application/json",
                **meta,
            },
        }

        return self.safe_json_stringify(response, escape_html=True)

    def create_error_response(
        self, message: str, code: str = "UNKNOWN_ERROR", details: Any = None
    ) -> str:
        """Create an error response"""
        response = {
            "success": False,
            "error": {
                "message": self.sanitize_string(message),
                "code": code,
                "timestamp": datetime.utcnow().isoformat(),
                "details": details,
            },
        }

        return self.safe_json_stringify(response, escape_html=True)


class SafeJSONEncoder(json.JSONEncoder):
    """Custom JSON encoder with additional safety features"""

    def encode(self, obj):
        """Override encode to add additional safety"""
        return super().encode(obj)

    def iterencode(self, obj, _one_shot=False):
        """Override iterencode for streaming safety"""
        return super().iterencode(obj, _one_shot)


# Global instance for easy use
json_safe = JsonSafeEncoder()

# Convenience functions
safe_json_parse = json_safe.safe_json_parse
safe_json_stringify = json_safe.safe_json_stringify
sanitize_string = json_safe.sanitize_string
create_api_response = json_safe.create_api_response
create_error_response = json_safe.create_error_response
